### PR TITLE
DDST-345: Create FOXML migration

### PR DESCRIPTION
**Pull request by chrismacdonaldw**

### CODE REVIEW

File: src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+      ['field_state_gov_classification'],
+      ['field_publication_volume_title'],
+      ['field_publication_chapter_number'],
+      ['field_publication_section'],
+      ['field_publication_url'],
+      ['field_origin_information', 3, 'values', '_field_date_created_single'],
+      ['field_origin_information', 3, 'values', '_field_date_created_start'],
+      ['field_origin_information', 3, 'values', '_field_date_created_end'],
+      ['field_origin_information', 3, 'values', '_field_date_created_assembled'],
+      ['field_origin_information', 3, 'values', 'field_date_created'],
```
Comment (jordandukart) [inline src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php:261]: Probably should nuke the backup field as well:

```
                'field_date_created_backup' => array (1) [
                    0 => array (5) [
                        'plugin' => string (4) "gate"
                        'source' => string (30) "@_field_date_created_assembled"
                        'use_as_key' => string (19) "@field_date_created"
                        'valid_keys' => array (1) [
                            0 => null
                        ]
                        'key_direction' => string (6) "unlock"
                    ]
                ]
```

File: src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    array_splice($process['field_publication_title'], 4, 0, [
+      [
+        'plugin' => 'callback',
+        'callable' => 'array_filter',
+      ],
+      [
+        'plugin' => 'null_coalesce',
+      ],
+    ]);
+
```
Comment (jordandukart) [inline src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php:233]: Looks like `field_publication_number` may need to be updated, from the dump:

```
 'field_publication_number' => array (8) [
        0 => array (5) [
            'plugin' => string (37) "dgi_migrate.process.xml.context_query"
            'missing_behavior' => string (12) "skip_process"
            'source' => string (11) "@_mods_node"
            'xpath' => string (12) "@_mods_xpath"
            'query' => string (61) "mods:relatedItem[@type="host"]/mods:titleInfo/mods:partNumber"
        ]
...
```
From the MMS: `mods:relatedItem[@type="host"]/mods:part/mods:detail[@type="issue"]/mods:number`.

File: src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    $process['field_remote_media_url'] = $process['field_ismn'];
+    $process['field_remote_media_url'][0]['query'] = 'mods:identifier[@displayLabel="remote media URL"]';
+
```
Comment (jordandukart) [inline src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php:236]: For both `field_extent_first_page` and `field_extent_last_page` the `@unit` attribute is missing on the XPath do we care?

```
'field_extent_first_page' => array (7) [
        0 => array (5) [
            'plugin' => string (37) "dgi_migrate.process.xml.context_query"
            'missing_behavior' => string (12) "skip_process"
            'source' => string (11) "@_mods_node"
            'xpath' => string (12) "@_mods_xpath"
            'query' => string (32) "mods:part/mods:extent/mods:start"
        ]
```
```
    'field_extent_last_page' => array (7) [
        0 => array (5) [
            'plugin' => string (37) "dgi_migrate.process.xml.context_query"
            'missing_behavior' => string (12) "skip_process"
            'source' => string (11) "@_mods_node"
            'xpath' => string (12) "@_mods_xpath"
            'query' => string (30) "mods:part/mods:extent/mods:end"
        ]
```

`mods:relatedItem/mods:part/mods:extent[@unit="pages"]/mods:start` and `mods:relatedItem/mods:part/mods:extent[@unit="pages"]/mods:end`

File: src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+      'plugin' => 'null_coalesce',
+    ];
+    $process['field_related_item_paragraph'][3]['values']['field_related_item_genre'][] = [
+      'plugin' => 'single_value',
+    ];
+    $process['field_related_item_paragraph'][3]['values']['field_related_item_genre'][] = [
+      'plugin' => 'callback',
+      'callable' => 'array_filter',
+    ];
+    $process['field_related_item_paragraph'][3]['values']['field_related_item_genre'][] = [
```
[the remark this answers is not in the document]
Comment (chrismacdonaldw) [reply src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php:188]: Genre has some updated authority sources, but the generic stub term anchor stuff should automatically handle any @authority stuff.

File: src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php (hunk 3 above)
Comment (chrismacdonaldw) [reply src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php:236]: Fixed.

File: src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php (hunk 2 above)
Comment (chrismacdonaldw) [reply src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php:233]: Fixed.

File: src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php (hunk 1 above)
Comment (chrismacdonaldw) [reply src/Plugin/dgi_migrate_alter/foxml/DgisNodesAlter.php:267]: Removed.

**Review by jordandukart — APPROVED** [review]

**Pull request merged**